### ISSUE TITLE
Bug in VQ function?

### ISSUE THREAD

**Issue by pfriesch** [ritheshkumar95/pytorch-vqvae#7]
Nice implementation of the VQ-Straight through function!

However, when looking at the autograd graph there is an edge that is breaking the separated gradients for the reconstruction loss and the VQ loss. So the reconstruction loss is also updating the embedding, which should not happen. I tried to figure out why that happens. My understanding of pytorch isn't that thorough though.  Do you might have an idea?

I marked the edge [here](https://github.com/ritheshkumar95/pytorch-vqvae/files/2031895/img_vqvae.gv.pdf).

**Comment by ritheshkumar95** [comment]
Hi @pfriesch, 

Thanks for pointing it out, in the previous version of the code, without the VQ-ST function (https://github.com/ritheshkumar95/pytorch-vqvae/blob/cde142670f701e783f29e9c815f390fc502532e8/vqvae.py#L75) we actually explicitly made sure that the reconstruction loss didn't update the codebook. 

We missed it when implementing the VQ-ST function. We're quickly fixing it. 

However, i'd like to point out that it didn't make much of a difference (reconstructions-wise, ELBO-wise), to do / do not update codebook with the reconstruction loss.

**Comment by tristandeleu** [comment]
This is something I did (incorrectly) on purpose, but you are absolutely right the paper explicitly says that the reconstruction loss does not update the embeddings. I fixed this.
Here is what the graph now looks like. This is a simplified version, where the left path is `loss_commit`, middle path is `loss_recons` and right path is `loss_vq`.
![image](https://user-images.githubusercontent.com/2018752/40445946-9644aae6-5e9b-11e8-97b0-59b5671bca16.png)
[Here](https://github.com/ritheshkumar95/pytorch-vqvae/files/2032503/full_graph.gv.pdf) is the full graph.

**Comment by pfriesch** [comment]
Cool! Thanks!

>However, i'd like to point out that it didn't make much of a difference (reconstructions-wise, ELBO-wise), to do / do not update codebook with the reconstruction loss.

Interesting, how does that work though? I am trying to wrap my head around that. That would mean the selected(closest) embeddings gets moved towards better fitting embeddings for the decoder, based on the reconstruction loss. Which in itself would not hurt the training of the vector quantized embedding and it might even speed up training. Would you agree with that intuition?

**Comment by ritheshkumar95** [comment]
Yes that is the same understanding that i had as well.
